Feature/crud documents

PAZderev opened the pull request.

PAZderev committed:
feat (Document) : Реализация CRUD операций для класса Document
 - Созданы контроллер, сервис, репозиторий для Document
 - Создан класс ApiError для возвращения унифицированной информации об ошибке
 - Создано исключение ResourceNotFoundException для 404 кода
 - Подправлены типы данных в БД

PAZderev committed:
fix (Document) : Переход на DTO Document в методах контроллера

PAZderev committed:
fix (Document) : Добавил создание AttributeValue вместе с созданием Document

PAZderev committed:
test (Document) : Unit и интеграционные тесты для CRUD Document

PAZderev committed:
fix () :fix conflicts

PAZderev committed:
feat (Document) : Разграничил DTO запроса и ответа в контроллере

PAZderev committed:
style () : checkstyle fix

PAZderev committed:
test (Document) : Учет перехода на DTO ответа в тестах

maksim25y commented on project/src/main/java/caselab/domain/entity/AttributeValueId.java:
```diff
@@ -0,0 +1,21 @@
+@Getter
+@Setter
+@EqualsAndHashCode
+public class AttributeValueId implements Serializable {
```
Не совсем понимаю зачем тут Serializable

PAZderev replied:
аннотация Embeddable требует

Класс не имеет своего собственного составного ключа, а использует два поля
```java
    private Long documentId;
    private Long attributeId;
```

PAZderev replied:
См. чуть ниже больше пояснил

maksim25y commented on project/src/main/java/caselab/controller/document/payload/DocumentResponseDTO.java:
```diff
@@ -0,0 +1,20 @@
+    private List<Long> applicationUserIds;
+    @ArraySchema(schema = @Schema(implementation = DocumentAttributeValueDTO.class))
+    private List<DocumentAttributeValueDTO> attributeValues;
+}
```
Как будто бы слишком много данных возвращает, 2 листа

maksim25y replied:
В payload ранее уже использовали рекорды, думаю есть смысл и тут их использовать

PAZderev replied:
Пока неизвестно вообще что пользователь должен получать, а что нет. Здесь надо от нужд клиента играть лучше
attributeValues 100% надо возвращать, насчет ID userов можно подумать



> Как будто бы слишком много данных возвращает, 2 листа

PAZderev replied:
> В payload ранее уже использовали рекорды, думаю есть смысл и тут их использовать

Сделал

Alexandr-Kokorin commented on project/src/main/java/caselab/configuration/WebConfig.java:
Зачем нужен данный конфиг, что он делает?

PAZderev replied:
Параметры пагинации по умолчанию задает

PAZderev replied:
Это задает параметры пагинации по умолчанию для всех методов где есть Pageable + уже локально в методах от этих настроек можно будет отступать

maksim25y commented on project/src/main/java/caselab/configuration/WebConfig.java:
```diff
@@ -0,0 +1,27 @@
+            DEFAULT_PAGE_SIZE)); // Параметры по умолчанию
+        resolvers.add(pageableResolver);
+    }
+}
```
Не совсем понимаю что это за класс и что он делает, случайно не захватил весь класс, речь о WebConfig в целом

PAZderev replied:
Параметры пагинации по умолчанию задает

PAZderev replied:
Это задает параметры пагинации по умолчанию для всех методов где есть Pageable + уже локально в методах от этих настроек можно будет отступать

Alexandr-Kokorin commented on project/src/main/java/caselab/controller/document/payload/DocumentAttributeValueDTO.java:
```diff
@@ -0,0 +1,16 @@
+    private Long id;
+    @Schema(description = "Значение атрибута, должно соответствовать типу атрибута", example = "Some value")
+    private String value;
+}
```
Кажется что лучше сделать Record и использовать Builder, раз нравятся сеттеры
```java
@Builder
public record DocumentAttributeValueDTO (
    @Schema(description = "ID атрибута", example = "1")
    long id;
    @Schema(description = "Значение атрибута, должно соответствовать типу атрибута", example = "Some value")
    String value;
) {
}
```

Alexandr-Kokorin commented on project/src/main/java/caselab/controller/document/payload/DocumentDTO.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.Getter;
+import lombok.Setter;
+
+
+@Getter
+@Setter
+public class DocumentDTO {
+    @Schema(description = "ID документа", example = "1")
+    private Long id;
+    @Schema(description = "ID Типа документа")
+    private Long documentTypeId;
+    @ArraySchema(schema = @Schema(description = "Список ID пользователей", example = "1"))
+    private List<Long> applicationUserIds;
+    @ArraySchema(schema = @Schema(implementation = DocumentAttributeValueDTO.class))
+    private List<DocumentAttributeValueDTO> attributeValues;
+}
```
Та же история с рекордом, а еще вообще кажется, что списки здесь не нужны, только первые два поля

PAZderev replied:
В моем понимании пользователь будет создавать документ, где будет выбирать тип документа и сразу к нему задавать значения

PAZderev replied:
перешел на record

Alexandr-Kokorin commented on project/src/main/java/caselab/controller/document/payload/DocumentResponseDTO.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.List;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class DocumentResponseDTO {
+    @Schema(description = "ID документа", example = "1")
+    private Long id;
+    @Schema(description = "ID Типа документа")
+    private Long documentTypeId;
+    @ArraySchema(schema = @Schema(description = "Список ID пользователей", example = "1"))
+    private List<Long> applicationUserIds;
+    @ArraySchema(schema = @Schema(implementation = DocumentAttributeValueDTO.class))
+    private List<DocumentAttributeValueDTO> attributeValues;
+}
```
Полная копия DocumentDTO, зачем он нужен? Или если хочется разделить на будущие изменения, то стоит называть не DTO, а Response и Request. Ну и опять таки рекорд

PAZderev replied:
Разбил на два класса -> Response и Request

Alexandr-Kokorin commented on project/src/main/java/caselab/domain/entity/AttributeValueId.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+
+@Embeddable
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+@EqualsAndHashCode
+public class AttributeValueId implements Serializable {
+    private Long documentId;
+    private Long attributeId;
+}
```
Зачем здесь сериализация? И опять кажется, что рекорд подойдет больше

PAZderev replied:
аннотация Embeddable требует

Класс не имеет своего собственного составного ключа, а использует два поля

    private Long documentId;
    private Long attributeId;

Alexandr-Kokorin replied:
Окей, тогда трогать не будем

PAZderev replied:
Вообще в небольшом приложении это можно и не указывать, код компилируется

Это больше следование спецификации JPA, она требует для такой аннотации Serializable

Проблемы возникнут, если мы реализуем систему кэша (Redis) либо микросервисы, там начнут исключения вылетать, поэтому лучше так

Alexandr-Kokorin commented on project/src/main/java/caselab/domain/entity/exception/ApiError.java:
Вот это логичнее будет смотреться на уровне контроллера, и опять таки рекорд

Alexandr-Kokorin commented on project/src/main/java/caselab/domain/entity/exception/ResourceNotFoundException.java:
Кажется тоже лучше вынести из entity, а может быть и из domain

PAZderev replied:
Удалил класс

Alexandr-Kokorin commented on project/src/main/java/caselab/service/document/DocumentService.java:
Очень длинные методы, пожалуйста разбей

Alexandr-Kokorin commented on project/src/test/java/caselab/controller/DocumentControllerTest.java:
Тесты окей, но мне кажется, можно добиться лучшего покрытия, хотелось бы увидеть соточку на контроллере и сервисе, это возможно, я с остальными добился)

Alexandr-Kokorin commented on project/src/test/java/caselab/service/DocumentServiceTest.java:
```diff
@@ -0,0 +1,202 @@
+        // Act
+        DocumentResponseDTO result = documentService.createDocument(documentDTO);
+
+        // Assert
+        assertNotNull(result);
+        assertNotNull(result.getId());
+        assertEquals(documentTypeId, result.getDocumentTypeId());
+        assertEquals(Arrays.asList(user1Id, user2Id), result.getApplicationUserIds());
+        assertEquals(attributeId, result.getAttributeValues().get(0).getId());
+        assertEquals("Test Value", result.getAttributeValues().get(0).getValue());
```
Можно объединить, пример: 
```java
assertAll("Grouped assertions for created document type",
                () -> assertEquals(actualDocumentType.id(), response.id()),
                () -> assertEquals(actualDocumentType.name(), payload.name()));
```

Alexandr-Kokorin commented on project/src/main/java/caselab/controller/document/DocumentController.java:
```diff
@@ -0,0 +1,58 @@
+        return documentService.getDocumentById(id);
+    }
+
+    @Operation(summary = "Получить страницу документов")
+    @GetMapping
+    public Page<DocumentResponseDTO> getAllDocuments(Pageable pageable) {
+        return documentService.getAllDocuments(pageable);
+    }
```
Прикол с Pageable не понял, зачем он?

PAZderev replied:
Пагинация

Alexandr-Kokorin replied:
Понял, но кажется, что проще вернуть список, а не страницу, но это скорее надо у фронта спрашивать

PAZderev replied:
Пагинация чаще серверная чем локальная, на каждую страницу посылается запрос, чтоб javascript не занимался ручной пагинацией

Alexandr-Kokorin reviewed (commented):
Я не уверен в правильности всех правок, предлагаю дождаться еще Лешу

DennKK commented on project/src/main/java/caselab/domain/entity/exception/ResourceNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class ResourceNotFoundException extends RuntimeException {
```
В проекте в DocumentTypesService используется NoSuchElementException. Возможно, имеет смысл использовать его вместо создания еще одного исключения, чтобы не создавать лишние сущности и не обрабатывать их потом в ControllerExceptionHandler.

Alexandr-Kokorin replied:
Согласен

ghost-of-endless replied:
От себя могу добавить, что `ResourceNotFoundException` - это ошибка org.springframework.web.cient, а лучше избегать наименований, который уже существуют в используемых фреймворках

PAZderev committed:
fix (Document) : Правки по комментариям
 - Переход на Record в DTO классах
 - Переход на NoSuchElementException вместо кастомного исключения
 - Рефактор методов DocumentService

DennKK commented on project/src/main/java/caselab/service/document/DocumentService.java:
```diff
@@ -0,0 +1,197 @@
+    public DocumentResponse getDocumentById(Long id) {
+        return toDTO(documentRepository.findById(id).orElseThrow(
+            () -> new NoSuchElementException(DOCUMENT_NOT_FOUND + id)
+        ));
```
У тебя довольно часто встречается эта конструкция. Стоит вынести её в отдельный метод. Например, можно создать метод getDocumentById, который будет искать документ по id, а если документ не найден, выбрасывать исключение

```
private Document getDocumentById(Long id) { 
    return documentRepository.findById(id)
        .orElseThrow(() -> new ResourceNotFoundException(DOCUMENT_NOT_FOUND + id));
}
```

И потом использовать в нужных местах

Alexandr-Kokorin commented:
Подтяни изменения из develop и добавь аннотацию авторизации к контроллерам

PAZderev committed:
fix (Document) : Правки по комментариям
 - Вынес выбрасывание исключения в отдельным метод в DocumentService
 - Добавил аннотацию в DocumentController

ghost-of-endless commented on project/src/main/java/caselab/service/document/DocumentService.java:
```diff
@@ -0,0 +1,200 @@
+    private AttributeValue createOrUpdateAttributeValue(Document document, DocumentAttributeValueDTO dto) {
+        Attribute attribute = attributeRepository.findById(dto.id())
+            .orElseThrow(() -> new NoSuchElementException(ATTRIBUTE_NOT_FOUND + dto.id()));
+        AttributeValue attributeValue = new AttributeValue();
+        attributeValue.setDocument(document);
+        attributeValue.setAttribute(attribute);
+        attributeValue.setAppValue(dto.value());
+        return attributeValue;
```
Может такое лучше на билдер переделать? Мне кажется, он бы тут красивей смотрелся

ghost-of-endless commented on project/src/main/java/caselab/service/document/DocumentService.java:
```diff
@@ -0,0 +1,200 @@
+        return document.getDocumentType() != null ? document.getDocumentType().getId() : null;
+    }
+
+    private List<Long> getApplicationUserIds(Document document) {
+        return document.getApplicationUsers() != null
+            ? document.getApplicationUsers().stream().map(ApplicationUser::getId).collect(Collectors.toList())
+            : null;
```
Вот тут прям Optional напрашивается, а то сложно будет потом по всему коду этот null ловить, да и читать тернарные операторы всё же тяжелее, чем методы Optional

ghost-of-endless commented on project/src/main/java/caselab/service/document/DocumentService.java:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    private List<DocumentAttributeValueDTO> getAttributeValuesDTO(Document document) {
+        return document.getAttributeValues() != null
+            ? document.getAttributeValues().stream()
+            .map(av -> new DocumentAttributeValueDTO(av.getAttribute().getId(), av.getAppValue()))
+            .collect(Collectors.toList())
+            : null;
+    }
```
Тут тоже хорошо бы увидеть Optional

GhostOfEndless commented on project/src/test/java/caselab/controller/DocumentControllerTest.java:
```diff
@@ -0,0 +1,157 @@
+    @MockBean
+    private SecurityFilterChain securityFilterChain;
+
+    private DocumentRequest documentRequest;
+    private DocumentResponse documentResponse;
```
А зачем их объявлять как поля класса, если ты их используешь как локальные переменные?

GhostOfEndless commented on project/src/test/java/caselab/controller/DocumentControllerTest.java:
```diff
@@ -0,0 +1,157 @@
+            .id(1L)
+            .value("Test Value")
+            .build();
+        documentRequest = DocumentRequest.builder()
```
Можно просто тут написать var, и поля не потребуются

ghost-of-endless commented on project/src/test/java/caselab/controller/DocumentControllerTest.java:
```diff
@@ -0,0 +1,157 @@
+        mockMvc.perform(post(DOCUMENT_URI)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(documentRequest)))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$.id").value(documentResponse.id()))
+            .andExpect(jsonPath("$.documentTypeId").value(documentResponse.documentTypeId()))
+            .andExpect(jsonPath("$.applicationUserIds[0]").value(1))
+            .andExpect(jsonPath("$.attributeValues[0].id").value(1))
+            .andExpect(jsonPath("$.attributeValues[0].value").value("Test Value"));
```
[Вот так ](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/test/web/servlet/ResultActions.html#andExpectAll(org.springframework.test.web.servlet.ResultMatcher...) "andExpectAll") было бы читабельней

ghost-of-endless reviewed (changes requested):
Вот null прям нужно убрать, а то потом намучаемся с ним

PAZderev committed:
fix (Document) : Правки по комментариям
 - Переход на Optional в DocumentService
 - Правка codestyle в тестах

ghost-of-endless commented on project/src/main/java/caselab/service/document/DocumentService.java:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    private Optional<List<DocumentAttributeValueDTO>> getAttributeValuesDTO(Document document) {
+        return Optional.ofNullable(document.getAttributeValues())
+            .map(attributeValues -> attributeValues.stream()
+                .map(av -> new DocumentAttributeValueDTO(av.getAttribute().getId(), av.getAppValue()))
+                .collect(Collectors.toList()));
+    }
```
Во, так намного лучше)

ghost-of-endless commented on project/src/test/java/caselab/controller/DocumentControllerTest.java:
```diff
@@ -0,0 +1,157 @@
+        mockMvc.perform(post(DOCUMENT_URI)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(documentRequest)))
+            .andExpectAll(status().isOk(),
+            jsonPath("$.id").value(documentResponse.id()),
+            jsonPath("$.documentTypeId").value(documentResponse.documentTypeId()),
+            jsonPath("$.applicationUserIds[0]").value(1),
+            jsonPath("$.attributeValues[0].id").value(1),
+            jsonPath("$.attributeValues[0].value").value("Test Value"));
```
Только `Ctrl+Alt+Shift+L` прожимай, чтобы форматировалось

PAZderev replied:
В будущей фиче прожму , чтоб ревью не слетело щас
